experimentation/src/main/java/com/automattic/android/experimentation/remote/ExperimentRestClient.kt

```diff
@@ -35,18 +35,24 @@ internal class ExperimentRestClient(
             .build()
 
         return withContext(dispatcher) {
-            okHttpClient.newCall(request).execute().use { response ->
-                if (!response.isSuccessful) {
-                    Result.failure(IOException("Unexpected code $response"))
-                } else {
-                    runCatching {
-                        val dto = jsonAdapter.fromJson(response.body!!.source())!!
-                        dto.toAssignments(
-                            fetchedAt = clock.currentTimeSeconds(),
-                            anonymousId = anonymousId,
-                        )
+            try {
```

Comment:
I've decided to go with a `try/catch` as this request is short-lived so I don't think we have to support graceful cancellation of this coroutine.

Reply:
I'm not sure I follow. `try/catch` and `runCatching` behave similarly and don't interact with the suspension mechanism. The main difference is that `try/catch` allows selective exception handling, whereas `runCatching` catches all exceptions, including `CancellationException`, which must be re-thrown. However, these two blocks are functionally equivalent:

```kotlin
suspend fun foo(): String = TODO()

val value = try {
  foo()
} catch (e: Throwable) {
  if (e is CancellationException) throw e
  null
}

val value = runCatching { foo() }
  .getOrElse { e ->
    if (e is CancellationException) throw e
    null
  }
```

In our case, though, it doesn’t matter because we aren't calling any suspending functions. `Call.execute()`, `JsonAdapter.fromJson()`, and `ResponseBody.source()` are all blocking, and we don’t invoke `yield()`, so there's no opportunity for coroutine cancellation during execution.

Reply:
> try/catch and runCatching behave similarly

I didn't mean `runCatching`. I rather meant `suspendCancellableCoroutine` idea you suggested on Slack.

> so there's no opportunity for coroutine cancellation during execution.

In the current shape - yes. But if I introduced `suspendCancellableCoroutine`, it'd look similar to

```kotlin
 suspendCancellableCoroutine { continuation ->
            val call = okHttpClient.newCall(request)
            continuation.invokeOnCancellation { call.cancel() }

            call.enqueue(object : Callback {
```

which would be cancellable.

Reply:
Right, I didn't get that from the message. Thanks for clarifying.